### PR TITLE
Add snap-layouts support to the Terminal

### LINKED ISSUE #8587: Title bar bug - Minimize button
# Environment

```none
Windows build number: Microsoft Windows NT 10.0.19042.0
Windows Terminal version (if applicable): 1.5.3242.0
```

# Steps to reproduce

1. Open Windows Terminal.
2. Click on minimize.
3. Put your cursor on the Terminal icon in the taskbar to get the application's preview.
4. Hover your cursor on the preview. You'll see that the minimize button is displayed as clicked and not released.

# Expected behavior

Minimize button should be displayed normally.

# Actual behavior

Minimize button is being displayed as clicked but not released.

### LINKED ISSUE #9443: Snap Layouts Support: WM_NCHITTEST should return HTMAXBUTTON, HTMINBUTTON, ... for caption buttons
Right now, our implementation of WM_NCHITTEST only covers the "caption" portion of the drag bar. To properly support accessibility tools, we'll need to return more than HTCAPTION.

This may require some rearchitecting to make sure the drag bar can pass input to the XAML input site.

### LINKED ISSUE #9447: Bracketed paste works strangely in WSLv1 with ZSH/oh-my-zsh shell
# Environment

```none
Platform      : Win32NT
ServicePack   :
Version       : 10.0.19041.0
VersionString : Microsoft Windows NT 10.0.19041.0

Windows Terminal version (if applicable):
Preview 1.7.572.0

WSLv1 running Fedora 33 and Ubuntu 20.04 with zsh/oh-my-zsh shell
PowerShell Core 7.1.2
```

# Steps to reproduce

Type some text and then paste whatever in the terminal.

# Expected behavior

Paste should work as in PowerShell - quickly with no overwriting.

# Actual behavior

Paste starts with overwriting command (though re-positions itself later), slowly adding characters, and leaves pasted text highlighted (until another key is pressed) when completed (left tab is PowerShell, right is Fedora on WSLv1 with oh-my-zsh).

paste_pwsh-vs-wsl

I have been checking other scenarios:

- PowerShell Windows, PowerShell Core, cmd - ok
- when I run the same WSLv1 but with bash as a shell - ok
- when I run WSL/oh-my-zsh session using Windows Terminal stable 1.6.10571.0 - ok (understandable - no bracketed paste there)
- when I paste text in python/ipython/vim from WSL/oh-my-zsh session - ok
- when I ssh from the WSL/oh-my-zsh session to a remote server running bash - ok
- when I ssh from the WSL/oh-my-zsh session to a remote server running similar zsh/oh-my-zsh combo - ~ok (paste works in a similar way - command will get overwritten first, but paste is very quick and text is not highlighted)
- when I run WSL with ZSH shell but without oh-my-zsh - ~ok (paste is quick, does not overwrite the command, but stays highlighted until another key is pressed)
- when I run WSL with ZSH shell with oh-my-zsh but disabled plugins (zsh-autosuggestions) - ~ok (paste is quick but delayed, does not overwrite the command, stays highlighted until another key is pressed)

From above, it looks that the problem may be related to zsh plugins, but I wonder, whether you could confirm whether this is or not Terminal's fault?

### PR DESCRIPTION
**Pull request by zadjii-msft**
Adds snap layout support to the Terminal's maximize button. This PR is
full of BODGY, so brace yourselves.

Big thanks to Chris Swan in #11134 for building the prototype.
I don't believe this solves #8795, because XAML islands can't get
nchittest messages

- The window procedure for the drag bar forwards clicks on its client
  area to its parent as non-client clicks.
- BODGY: It also _manually_ handles the caption buttons. They exist in
  the titlebar, and work reasonably well with just XAML, if the drag bar
  isn't covering them.
- However, to get snap layout support, we need to actually return
  `HTMAXBUTTON` where the maximize button is. If the drag bar doesn't
  cover the caption buttons, then the core input site (which takes up
  the entirety of the XAML island) will steal the `WM_NCHITTEST` before
  we get a chance to handle it.
- So, the drag bar covers the caption buttons, and manually handles
  hovering and pressing them when needed. This gives the impression that
  they're getting input as they normally would, even if they're not
  _really_ getting input via XAML.
- We also need to manually display the button tooltips now, because XAML
  doesn't know when they've been hovered for long enough. Hence, the
  `_displayToolTip` `ThrottledFuncTrailing`

## Validation
Minimized, maximized, restored down, hovered the buttons slowly, moved
the mouse over them quickly, they feel the same as before. But now with
snap layouts appearing.

## TODO!
* [x] I'm working on getting the ToolTips on the caption buttons back. Alas, I needed a demo of this _today_, so I'll fix that tomorrow morning.
* [x] mild concern: I should probably test Win 10 to make sure there wasn't weird changes to the message loop in win11 that means this is broken on win10.
* [x] I think I used the wrong issue number for tons of my comments throughout this PR. Double check that. Should be #9443, not #9447. 

Closes #9443
I thought this took care of #8587 ~as a bonus, because I was here, and the fix is _now_ trivial~, but looking at the latest commit that regressed.

Co-authored-by: Chris Swan <chswan@microsoft.com>

### CODE REVIEW

**Comment by miniksa** [comment]
>  This PR is full of BODGY, so brace yourselves.

Oh man am I excited.

**Comment by miniksa** [comment]
> * [x]  Closes [Bracketed paste works strangely in WSLv1 with ZSH/oh-my-zsh shell #9447](https://github.com/microsoft/terminal/issues/9447)

Wait what?

File: src/cascadia/TerminalApp/MinMaxCloseControl.xaml [hunk 1]
```diff
@@ -309,5 +314,10 @@
                 <x:String x:Key="CaptionButtonPath">M 0 0 L 10 10 M 10 0 L 0 10</x:String>
             </ResourceDictionary>
         </Button.Resources>
+        <ToolTipService.ToolTip>
+            <ToolTip>
+                <TextBlock x:Uid="WindowCloseButtonToolTip" />
```
Comment (miniksa) [inline src/cascadia/TerminalApp/MinMaxCloseControl.xaml:319]: Why is the Maximize one the odd duck out using a <Run> here that is named?

File: src/cascadia/TerminalApp/TitlebarControl.cpp [hunk 2]
```diff
@@ -24,6 +23,12 @@ namespace winrt::TerminalApp::implementation
         MinMaxCloseControl().CloseClick({ this, &TitlebarControl::Close_Click });
     }
 
+    double TitlebarControl::CaptionButtonWidth()
+    {
+        static double width{ MinMaxCloseControl().ActualWidth() / 3.0 };
```
Comment (miniksa) [inline src/cascadia/TerminalApp/TitlebarControl.cpp:28]: Note why 3. Just wait until someone designs a fourth button up there. :|

File: src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp [hunk 3]
```diff
@@ -87,51 +87,178 @@ void NonClientIslandWindow::MakeWindow() noexcept
...
-// - The window procedure for the drag bar forwards clicks on its client area to its parent as non-client clicks.
-LRESULT NonClientIslandWindow::_InputSinkMessageHandler(UINT const message, WPARAM const wparam, LPARAM const lparam) noexcept
+LRESULT NonClientIslandWindow::_dragBarNcHitTest(const til::point& pointer)
 {
-    std::optional<UINT> nonClientMessage{ std::nullopt };
+    RECT rcParent = GetWindowRect();
+    // The size of the buttons doesn't change over the life of the application.
+    static const auto buttonWidthInDips{ _titlebar.CaptionButtonWidth() };
+
+    // However, the dPI scaling might, so get the updated size of the buttons in pixels
```
Comment (miniksa) [inline src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp:96]:
```suggestion
    // However, the DPI scaling might, so get the updated size of the buttons in pixels

```

**Review by miniksa — COMMENTED** [review]
Does Alt+Space also still work for the system menu? This was only positioning for mouse events right? I don't think the mouse-variant of the system menu in the top left ever worked for us (though we could make it work maybe by returning `HTSYSMENU`, eh?

File: src/cascadia/TerminalApp/MinMaxCloseControl.xaml (hunk 1 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/MinMaxCloseControl.xaml:319]: The Maximize button needs to have it's tooltip text changed in code-behind depending on if it's "Maximize" or "Restore down"

**Review by DHowett — CHANGES_REQUESTED** [review]
Looks like I never submitted this.

Notes from use:
* click and hold a caption button: it unhighlights itself when the tooltip comes up
* the caption buttons no longer animate state transition
* _the top resize border doesn't work_ (!)

**Comment by DHowett** [comment]
> System menu

Doesn't work when you right-click the title bar.

Also, weird thing happening on my 150% scale display:

wt-bug-caption-buttons

**Comment by zadjii-msft** [comment]
> Also, weird thing happening on my 150% scale display:

This is why I need a high DPI display to test, instead of just blindly writing the code 😅

File: src/cascadia/TerminalApp/MinMaxCloseControl.xaml (hunk 1 above)
Comment (miniksa) [reply src/cascadia/TerminalApp/MinMaxCloseControl.xaml:319]: Yuck OK.

**Comment by zadjii-msft** [comment]
* [ ] "Also, weird thing happening on my 150% scale display"
  * fixed in  37c4cbee7
* [ ] click and hold a caption button: it unhighlights itself when the tooltip comes up
  * Weird, shouldn't be too bad...
* [ ] the caption buttons no longer animate state transition
  * Weird, shouldn't be too bad...
* [ ] the top resize border doesn't work (!)
  * _F E A R_ 

will update as needed

**Comment by DHowett** [comment]
This is also going out in a selfhost build this evening. :smile:

**Comment by DHowett** [comment]
Alright, I've only caught one issue with this and that's: when you click down on close, move to maximize, then release, it _performs maximize_ instead of sinking the event.
If it's an easy fix, I'd like it, and if not... we can fix it later?

**Comment by DHowett** [comment]
before you merge this, drive those todos to 0 and delete them? :smile:

File: src/cascadia/TerminalApp/Resources/en-US/Resources.resw [hunk 4]
```diff
@@ -411,6 +411,9 @@
   <data name="WindowCloseButton.[using:Windows.UI.Xaml.Controls]ToolTipService.ToolTip" xml:space="preserve">
```
Comment (DHowett) [inline src/cascadia/TerminalApp/Resources/en-US/Resources.resw:411]: too bad we can't reuse the existing tooltip machinery

**Review by DHowett — APPROVED** [review]

**Comment by miniksa** [comment]
> before you merge this, drive those todos to 0 and delete them? 😄

This is why I didn't hit approve here... waiting for Mike's input

**Review by miniksa — APPROVED** [review]

**Comment by DHowett** [comment]
Moved from body:

gh-9447-prototype-000

**Comment by augustresende** [comment]
If anyone is interested, snap layouts does not work in flutter desktop apps if title bar is customized/disabled:

https://github.com/flutter/flutter/issues/156791

**Pull request merged**